### ISSUE TITLE
Implement "autoloading" source locator

### ISSUE THREAD

**Issue by asgrim** [Roave/BetterReflection#36]
In order to automatically reflect using the exact same API as the core `Reflection*` classes, we need to be able to essentially ignore the whole `SourceLocator` concept, despite it's flexibility.

So, I think it's possible to implement a new `AutoloadingSourceLocator` that takes no constructor dependencies, which, funnily enough, would probably need to rely on the internal `ReflectionClass` API, e.g.:

```php
// if it's a class we're loading...
$reflectionClass = new \ReflectionClass($identifier->getName());
$filename = $reflectionClass->getFileName();
$source = file_get_contents($filename);
// use these to make the LocatedSource...
```

Additionally, we'd need to create a "static" reflector, which would be something like:

```php
class StaticReflector
{
    public static function reflect(...)
    {
        return (new Reflector(new AutoloadingSourceLocator()))->reflect($identifier);
    }
}
```

Then we can use this reflector in the constructors or something to reflect (still a bit hazy on how to actually implement this). The downside of this is it means we do have to LOAD the classes. So I also propose making a `LoadSafeSourceLocator` interface, which doesn't have any methods, but purely indicates that it "by design" does NOT load up classes into memory. Then in the `..\Reflector\Generic` class, we can check if `$reflection implements LoadSafeSourceLocator`, and ensure that the class is not loaded beforehand, and is not loaded after reflection, to ensure the contract is obeyed.

### TL;DR

* Create new `AutoloadingSourceLocator` that uses core refleciton API to locate source code filename
* Cleanup constructors so they are empty
* Create `StaticReflector` that uses `AutoloadingSourceLocator`
* Update `Reflection*` classes to use `AutoloadingSourceLocator`
* Create `LoadSafeSourceLocator` interface, any source locator that implements this interface MUST ensure the class does not get loaded ever. These would be the `Composer`, `SingleFile` and `String` source locators, as these do not load file. `AutoloadingSourceLocator` would load the file, so MUST NOT implement the `LoadSafeSourceLocator` interface.
* This allows us to do the following (i.e. replicate almost _exactly_ the core reflection API instantiation technique):

```php
$reflectionClass = new \BetterReflection\Reflection\ReflectionClass('My\Foo\Bar');
```

**Comment by Ocramius** [comment]
> we'd need to create a "static" reflector

I don't see a need for it

 > Then we can use this reflector in the constructors or something to reflect (still a bit hazy on how to actually implement this). 

Probably simply rely on the internal `ReflectionClass` and use `class_exists($class, false)` first (to avoid autoloading, if not an internal class). Then use `ReflectionClass#isInternal()`

 > The downside of this is it means we do have to LOAD the classes. 

As shown above, can be avoided. The locator would simply fail for non-internal classes, or classes that aren't already loaded.

**Comment by asgrim** [comment]
Apologies - original intent was not clear by misusing "internal" - I had intended this to be a regular "autoloading" source locator - I have amended the wording above. (btw, for reference, the ability to reflect on internal stuff would be in #38 )

**Comment by Ocramius** [comment]
As per discussion, one way to *try* implementing this (without any knowledge about the autoloader in use) is to override and restore the stream wrapper for `file://`. I am not sure if that is possible, but it can prevent the autoloader from requiring a file. The wrapper would pass the contents of the file to the reflector, and then return an empty string to the autoloader, effectively causing the autoloader to not find the class.

**Comment by GeeH** [comment]
The reason we need this is to keep parity with the native `\ReflectionClass`. According to the documentation the  `getInterfaces()` method will return:

> An associative array of interfaces, with keys as interface names and the array values as ReflectionClass objects.

This means we need to iterate of the `implements` key in the AST and recursively build the interface `ReflectionClass` objects, wether lazily or on construct is an implementation detail.

**Comment by asgrim** [comment]
On further thoughts/discussions, I'm scrapping the "LoadSafe" stuff, and just keeping basically what is implemented in #41.
